midterm-2023-11-06/code/5029-angel-nineski/script.ts

```diff
@@ -0,0 +1,184 @@
+        const lastAlbumYear = Math.max(...band.albums.map(album => album.year));
+
+        bandElement.innerHTML = `
+            <div class="band-data">${band.id}</div>
+            <div class "band-data">${band.name}</div>
+            <div class="band-data">${band.formed}</div>
+            <div class="band-data">${location[0]}</div>
+            <div class="band-data">${country}</div>
+            <div class="band-data">${band.members.join(", ")}</div>
+            <div class="band-data">${albums}</div>
+            <div class="band-data">${firstAlbumYear}</div>
+            <div class="band-data">${lastAlbumYear}</div>
+            <div class="band-data">${firstAlbumYear} - ${lastAlbumYear}</div>
+        `;
+
+        bandContainer.appendChild(bandElement);
+    });
+
+    const headers = document.querySelectorAll(".band-header");
+    headers.forEach((header) => {
+        header.addEventListener("click", () => {
+            const sortBy = header.getAttribute("data-sort-by");
+            if (sortBy) {
+                sortTable(bands, sortBy);
+                displayBands(bands);
+            }
+        });
+    });
+}
+
+// Function to sort the table data
+function sortTable(bands: Band[], sortBy: string) {
+    bands.sort((a, b) => {
+        let aValue, bValue: number | string;
+
+        switch (sortBy) {
+            case "id":
+                aValue = a.id;
+                bValue = b.id;
+                break;
+            case "name":
+            case "location":
+            case "country":
+                aValue = a[sortBy];
+                bValue = b[sortBy];
+                break;
+            case "formed":
+            case "firstAlbumYear":
+            case "lastAlbumYear":
+                aValue = a[sortBy];
+                bValue = b[sortBy];
+                break;
+            case "yearsActive":
+                const aStartYear = a.albums[0].year;
+                const aEndYear = a.albums[a.albums.length - 1].year;
+                const bStartYear = b.albums[0].year;
+                const bEndYear = b.albums[b.albums.length - 1].year;
+                return aStartYear - bStartYear || aEndYear - bEndYear;
+            default:
+                aValue = 0;
+                bValue = 0;
+        }
+
+        if (typeof aValue === 'string' && typeof bValue === 'string') {
+            return (aValue as string).localeCompare(bValue as string);
+        } else if (typeof aValue === 'number' && typeof bValue === 'number') {
+            return (aValue as number) - (bValue as number);
+        } else {
+            return 0;
+        }
+    });
+}
+
+// Function to populate filter options
+function populateFilterOptions(data: any) {
+    const countryFilter = document.getElementById("country-filter");
+    const genreFilter = document.getElementById("genre-filter");
+
+    const countries = new Set<string>();
+    const genres = new Set<string>();
+
+    data.metalBands.forEach((band: Band) => {
```

Comment:
Not totally sure that this works, as the bands do not have a `country` property, we're calculating and displaying that in runtime.